# 1.Load data

Please download the dataset ner_dataset.csv from https://www.kaggle.com/abhinavwalia95/entity-annotated-corpus

In [76]:
import pandas as pd

In [77]:
data = pd.read_csv('./kaggleData/ner_dataset.csv', encoding= 'unicode_escape')
data

Sentence #           Word  POS Tag
0        Sentence: 1      Thousands  NNS   O
1                NaN             of   IN   O
2                NaN  demonstrators  NNS   O
3                NaN           have  VBP   O
4                NaN        marched  VBN   O
...              ...            ...  ...  ..
1048570          NaN           they  PRP   O
1048571          NaN      responded  VBD   O
1048572          NaN             to   TO   O
1048573          NaN            the   DT   O
1048574          NaN         attack   NN   O

[1048575 rows x 4 columns]

## 2. Extract mappings required for the neural network
To train a neural network, we will use two mappings as given below. The neural network will only take integers as input. So lets convert all the unique tokens in the corpus to its respective index.
- {token} to {token id}: address the row in embeddings matrix for the current token.
- {tag} to {tag id}: one-hot ground truth probability distribution vectors for computing the loss at the output of the network.

In [78]:
from itertools import chain
def get_dict_map(data, token_or_tag):
    tok2idx = {}
    idx2tok = {}
    
    if token_or_tag == 'token':
        vocab = list(set(data['Word'].to_list()))
    else:
        vocab = list(set(data['Tag'].to_list()))
    
    idx2tok = {idx:tok for  idx, tok in enumerate(vocab)}
    tok2idx = {tok:idx for  idx, tok in enumerate(vocab)}
    return tok2idx, idx2tok


token2idx, idx2token = get_dict_map(data, 'token')
tag2idx, idx2tag = get_dict_map(data, 'tag')

In [79]:
data['Word_idx'] = data['Word'].map(token2idx)
data['Tag_idx'] = data['Tag'].map(tag2idx) 

In [80]:
data

Sentence #           Word  POS Tag  Word_idx  Tag_idx
0        Sentence: 1      Thousands  NNS   O     24686        8
1                NaN             of   IN   O     29967        8
2                NaN  demonstrators  NNS   O     31412        8
3                NaN           have  VBP   O     31399        8
4                NaN        marched  VBN   O     13154        8
...              ...            ...  ...  ..       ...      ...
1048570          NaN           they  PRP   O      3647        8
1048571          NaN      responded  VBD   O     25704        8
1048572          NaN             to   TO   O     25924        8
1048573          NaN            the   DT   O     10516        8
1048574          NaN         attack   NN   O     17925        8

[1048575 rows x 6 columns]

## 3. Transform columns to extract sequential data
Next, lets fill NaN in 'sentence #' column using method ffill in fillna. Thereafter groupby on the sentence column to get a list of tokens and tags for each sentence.

In [81]:
data_fillna = data.fillna(method='ffill', axis=0)
data_group = data_fillna.groupby(['Sentence #'],as_index=False
                                )['Word', 'POS', 'Tag', 'Word_idx', 'Tag_idx'].agg(lambda x: list(x))

#data_fillna
data_group

<ipython-input-81-a7bf3c314e32>:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  data_group = data_fillna.groupby(['Sentence #'],as_index=False


Sentence #                                               Word  \
0          Sentence: 1  [Thousands, of, demonstrators, have, marched, ...   
1         Sentence: 10  [Iranian, officials, say, they, expect, to, ge...   
2        Sentence: 100  [Helicopter, gunships, Saturday, pounded, mili...   
3       Sentence: 1000  [They, left, after, a, tense, hour-long, stand...   
4      Sentence: 10000  [U.N., relief, coordinator, Jan, Egeland, said...   
...                ...                                                ...   
47954   Sentence: 9995  [Opposition, leader, Mir, Hossein, Mousavi, ha...   
47955   Sentence: 9996  [On, Thursday, ,, Iranian, state, media, publi...   
47956   Sentence: 9997  [Following, Iran, 's, disputed, June, 12, elec...   
47957   Sentence: 9998  [Since, then, ,, authorities, have, held, publ...   
47958   Sentence: 9999  [The, United, Nations, is, praising, the, use,...   

                                                     POS  \
0      [NNS, IN, NNS, VBP, VBN, IN, NNP, TO, VB, DT, ...   
1      [JJ, NNS, VBP, PRP, VBP, TO, VB, NN, TO, JJ, J...   
2      [NN, NNS, NNP, VBD, JJ, NNS, IN, DT, NNP, JJ, ...   
3         [PRP, VBD, IN, DT, NN, JJ, NN, IN, NN, NNS, .]   
4      [NNP, NN, NN, NNP, NNP, VBD, NNP, ,, NNP, ,, J...   
...                                                  ...   
47954  [NNP, NN, NNP, NNP, NNP, VBZ, VBN, PRP, VBZ, T...   
47955  [IN, NNP, ,, JJ, NN, NNS, VBN, DT, NN, IN, DT,...   
47956  [VBG, NNP, POS, JJ, NNP, CD, NNS, ,, NNS, NNS,...   
47957  [IN, RB, ,, NNS, VBP, VBN, JJ, NNS, IN, DT, VB...   
47958  [DT, NNP, NNP, VBZ, VBG, DT, NN, IN, JJ, NNS, ...   

                                                     Tag  \
0      [O, O, O, O, O, O, B-geo, O, O, O, O, O, B-geo...   
1      [B-gpe, O, O, O, O, O, O, O, O, O, O, O, O, O,...   
2      [O, O, B-tim, O, O, O, O, O, B-geo, O, O, O, O...   
3                      [O, O, O, O, O, O, O, O, O, O, O]   
4      [B-geo, O, O, B-per, I-per, O, B-tim, O, B-geo...   
...                                                  ...   
47954  [O, O, O, B-per, I-per, O, O, O, O, O, O, O, O...   
47955  [O, B-tim, O, B-gpe, O, O, O, O, O, O, O, O, B...   
47956  [O, B-geo, O, O, B-tim, I-tim, O, O, O, O, O, ...   
47957  [O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...   
47958  [O, B-org, I-org, O, O, O, O, O, O, O, O, O, O...   

                                                Word_idx  \
0      [24686, 29967, 31412, 31399, 13154, 20503, 264...   
1      [15935, 3563, 705, 3647, 24063, 25924, 1180, 2...   
2      [30370, 11407, 14776, 22738, 15936, 29053, 177...   
3      [1359, 26607, 9877, 25043, 32903, 18546, 8855,...   
4      [29394, 32344, 22279, 33769, 13501, 2263, 3450...   
...                                                  ...   
47954  [8744, 29659, 14517, 1688, 5585, 3537, 2263, 2...   
47955  [1221, 21334, 6880, 15935, 17262, 29503, 12378...   
47956  [18739, 6766, 2551, 7130, 34890, 29997, 2873, ...   
47957  [6814, 27694, 6880, 24548, 31399, 24283, 11289...   
47958  [34844, 27180, 10331, 33569, 24444, 10516, 119...   

                                                 Tag_idx  
0      [8, 8, 8, 8, 8, 8, 12, 8, 8, 8, 8, 8, 12, 8, 8...  
1      [10, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,...  
2      [8, 8, 15, 8, 8, 8, 8, 8, 12, 8, 8, 8, 8, 8, 1...  
3                      [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8]  
4      [12, 8, 8, 11, 3, 8, 15, 8, 12, 8, 10, 8, 10, ...  
...                                                  ...  
47954  [8, 8, 8, 11, 3, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,...  
47955  [8, 15, 8, 10, 8, 8, 8, 8, 8, 8, 8, 8, 13, 0, ...  
47956  [8, 12, 8, 8, 15, 1, 8, 8, 8, 8, 8, 8, 8, 8, 8...  
47957  [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, ...  
47958  [8, 13, 0, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,...  

[47959 rows x 6 columns]

## 4. Pad sequences and split the dataset into train, test
Padding: The LSTM layers accept sequences of same length only. Therefore we will want to transform our list of token_sequences ('Word_idx') which is lists of integers into a matrix of shape (token_sequences, max_len). We can use any length as max_len. In this project we will be using length of the longest sequence as max_len. The sequences that are shorter than max_len are padded with a specified value at the end.
Import required packages

In [82]:
from sklearn.model_selection import train_test_split
from keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical

In [83]:
def get_pad_train_test_val(data_group, data):
    n_token = len(list(set(data['Word'].to_list())))
    n_tag = len(list(set(data['Tag'].to_list())))
    print(n_token)
    print(n_tag)
    tokens = data_group['Word_idx'].tolist()
    maxlen = max([len(s) for s in tokens])
    print(maxlen)
    pad_tokens = pad_sequences(tokens, maxlen=maxlen, dtype='int32', padding='post', value= n_token - 1)
    print(pad_tokens)
    tags = data_group['Tag_idx'].tolist()
    pad_tags = pad_sequences(tags, maxlen=maxlen, dtype='int32', padding='post', value= tag2idx["O"])
    print(pad_tags)
    n_tags = len(tag2idx)
    print(n_tags)
    pad_tags = [to_categorical(i, num_classes=n_tags) for i in pad_tags]
    train_tokens, test_tokens, train_tags, test_tags = train_test_split(pad_tokens, pad_tags, test_size=0.1, train_size=0.9, random_state=2020)

    print(
        'train_tokens length:', len(train_tokens),
        '\ntest_tokens length:', len(test_tokens),
        '\ntrain_tags:', len(train_tags),
        '\ntest_tags:', len(test_tags)
    )
    
    return train_tokens, test_tokens, train_tags, test_tags

train_tokens, test_tokens, train_tags, test_tags = get_pad_train_test_val(data_group, data)

35178
17
104
[[24686 29967 31412 ... 35177 35177 35177]
 [15935  3563   705 ... 35177 35177 35177]
 [30370 11407 14776 ... 35177 35177 35177]
 ...
 [18739  6766  2551 ... 35177 35177 35177]
 [ 6814 27694  6880 ... 35177 35177 35177]
 [34844 27180 10331 ... 35177 35177 35177]]
[[ 8  8  8 ...  8  8  8]
 [10  8  8 ...  8  8  8]
 [ 8  8 15 ...  8  8  8]
 ...
 [ 8 12  8 ...  8  8  8]
 [ 8  8  8 ...  8  8  8]
 [ 8 13  0 ...  8  8  8]]
17
train_tokens length: 43163 
test_tokens length: 4796 
train_tags: 43163 
test_tags: 4796


In [68]:
train_tags[1][0]

array([0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
      dtype=float32)

In [69]:
for token, tag in zip(train_tokens[0], train_tags[0]):
    print('%s\t%s' % (token, tag))

12848	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
705	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
30195	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
33569	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
35032	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
10559	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
25924	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
27284	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
12739	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
20981	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
6880	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
29534	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
3647	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
32299	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
20357	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
3932	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
15249	[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
66

## 5. Build Model Layout

Lets go through the process of building a neural network model with lstm layers. Please compare the layers brief and model plot given below to get a better understanding of the layers, input and output dimensions. We are building a simple model with 4 layers.

- **Layer 1 - Embedding layer** : We will feed the padded sequences of equal length (104) to the embedding layer. Once the network has been trained, each token will get transformed into a vector of n dimensions. We have chosen the n dimensions to be (64). 


These are the dimensions (?, 104, 64) plotted in the model plot for input layer and embedding layer. The ? or None in the dimension specifies batches, when it is None or ? the model can take any batch size.

- **Layer 2 - Bidirectional LSTM** : Bidirectional lstm takes a recurrent layer (e.g. the first LSTM layer) as an argument. This layer takes the output from the previous embedding layer (104, 64). It also allows you to specify the merge mode, that is how the forward and backward outputs should be combined before being passed on to the next layer. The default mode is to concatenate, where the outputs are concatenated together, providing double the number of outputs to the next layer, in our case its 128(64 * 2).


The output dimension of the bidirectional lstm layer (?, 104, 128) becomes the input dimension of the next lstm layer.

- **Layer 3 - LSTM Layer** : An LSTM network is a recurrent neural network that has LSTM cell blocks in place of our standard neural network layers. These cells have various components called the input gate, the forget gate and the output gate.


This layer takes the output dimension from the previous bidirectional lstm layer (?, 104, 128) and outputs (?, 104, 256)



- **Layer 4 - TimeDistributed  Layer** : We are dealing with Many to Many RNN Architecture where we expect output from every input sequence for example (a1 →b1, a2 →b2… an →bn) where a and b are inputs and outputs of every sequence. The TimeDistributeDense layers allow you to apply Dense(fully-connected) operation across every output over every time-step. If you don't use this, you would only have one final output.



This layer take the output dimension of the previous lstm layer (104, 256) and outputs the max sequence length (104) and max tags (17).

In [84]:
import numpy as np
import tensorflow
from tensorflow.keras import Sequential, Model, Input
from tensorflow.keras.layers import LSTM, Embedding, Dense, TimeDistributed, Dropout, Bidirectional
from tensorflow.keras.utils import plot_model

# import keras as k

In [85]:
from numpy.random import seed
seed(1)
tensorflow.random.set_seed(2)

In [86]:
input_dim = len(list(set(data['Word'].to_list())))+1
output_dim = 32
input_length = max([len(s) for s in data_group['Word_idx'].tolist()])
n_tags = len(tag2idx)
print('input_dim: ', input_dim, '\noutput_dim: ', output_dim, '\ninput_length: ', input_length, '\nn_tags: ', n_tags)

input_dim:  35179 
output_dim:  32 
input_length:  104 
n_tags:  17


In [87]:
def get_bilstm_lstm_model():
    model = Sequential()

    # Add Embedding layer
    model.add(Embedding(input_dim=input_dim, output_dim=output_dim, input_length=input_length))

    # Add bidirectional LSTM
    model.add(Bidirectional(LSTM(units=output_dim, return_sequences=True, dropout=0.2, recurrent_dropout=0.2), merge_mode = 'concat'))

    # Add LSTM
    model.add(LSTM(units=output_dim, return_sequences=True, dropout=0.5, recurrent_dropout=0.5))

    # Add timeDistributed Layer
    model.add(TimeDistributed(Dense(n_tags, activation="relu")))

    #Optimiser 
    # adam = k.optimizers.Adam(lr=0.0005, beta_1=0.9, beta_2=0.999)

    # Compile model
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
    model.summary()
    
    return model

In [90]:
def train_model(X, y, model):
    loss = list()
    for _ in range(5):
        # fit model for one epoch on this sequence
        hist = model.fit(X, y, batch_size=1000, verbose=1, epochs=1, validation_split=0.2)
        loss.append(hist.history['loss'][0])
    return loss
model_bilstm_lstm = get_bilstm_lstm_model()
plot_model(model_bilstm_lstm)

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_3 (Embedding)     (None, 104, 32)           1125728   
                                                                 
 bidirectional_3 (Bidirectio  (None, 104, 64)          16640     
 nal)                                                            
                                                                 
 lstm_7 (LSTM)               (None, 104, 32)           12416     
                                                                 
 time_distributed_3 (TimeDis  (None, 104, 17)          561       
 tributed)                                                       
                                                                 
Total params: 1,155,345
Trainable params: 1,155,345
Non-trainable params: 0
_________________________________________________________________
('You must install pydot (`pip install pydot

In [91]:
results = pd.DataFrame()
results['with_add_lstm'] = train_model(train_tokens, np.array(train_tags), model_bilstm_lstm)

35/35 [==============================] - 109s 3s/step - loss: 0.2321 - accuracy: 0.9677 - val_loss: 0.2126 - val_accuracy: 0.9679


In [100]:
test_tags[0][3]

array([0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
      dtype=float32)

In [101]:
predict_tags = model_bilstm_lstm.predict(test_tokens)

In [104]:
predict_tags[0]

array([[0.00026819, 0.        , 0.        , ..., 0.        , 0.00569805,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.0100915 ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.0157113 ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]], dtype=float32)